File: 028_BinarySearch.cpp
// Topic: Binary Search
// Description: Search target value in sorted array efficiently using divide-and-conquer.
// Approach: Standard iterative binary search returning index or -1.

#include <bits/stdc++.h>
using namespace std;

int binarySearch(const vector<int>& a, int target)
{
    int l = 0, r = (int)a.size() - 1;
    while (l <= r)
    {
        int mid = l + ((r - l) >> 1);
        if (a[mid] == target)
            return mid;
        if (a[mid] < target)
            l = mid + 1;
        else
            r = mid - 1;
    }
    return -1;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    vector<pair<vector<int>, int>> tests = {{{-1, 0, 3, 5, 9, 12}, 9}, {{1, 2, 3, 4, 5}, 2}};
    for (size_t i = 0; i < tests.size(); ++i)
    {
        cout << "Test " << (i + 1) << ": target=" << tests[i].second << " -> "
             << binarySearch(tests[i].first, tests[i].second) << "\n";
    }
}

/*
Time Complexity: O(log n)
Space Complexity: O(1)
*/

File: 028_BinarySearch.py
# Topic: Binary Search
# Description: Search target value in sorted array efficiently using divide-and-conquer.
# Approach: Standard iterative binary search returning index or -1.

from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                return mid
            if nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return -1


if __name__ == "__main__":
    tests = [([-1, 0, 3, 5, 9, 12], 9), ([1, 2, 3, 4, 5], 2)]
    sol = Solution()
    for i, (nums, target) in enumerate(tests):
        result = sol.search(nums, target)
        print(f"Test {i + 1}: target={target} -> {result}")

"""
Time Complexity: O(log n)
Space Complexity: O(1)
"""
